subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/AutoInstalledInstallationSupplier.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain.internal;
+
+import org.gradle.jvm.toolchain.install.internal.JdkCacheDirectory;
+
+import java.io.File;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class AutoInstalledInstallationSupplier implements InstallationSupplier {
```

Comment:
Would it make sense to make this a "cache" and then have it trigger the download if the cache entry is not there. Instead of wiring in the downloader separately in the query service?

I also have the feeling we should be able to use some of our existing infrastructure to make it a "stable" cache. So that we notice if something has been modified in the extracted directory and also to address the issue of parllale access (triggering potential parallel download - see other TODO and comment about that).

Unfortunately I am not sure which part of the caching infrastructure fits here from the top of my head.

Reply:
I was thinking about that approach initially. Two reasons that made me not do it:
* it requires suppliers to know about the spec to know what to download. That means we can't cache all the other suppliers anymore
* it requires a concept of ordering between suppliers as the download suppliers should always be the last ones

This led me to separate the concerns of detecting and downloading toolchains